### DISCUSSION TITLE
How the sample works?

### DISCUSSION THREAD

**Discussion by dfguerrerom** [torchgeo/torchgeo#659] [Q&A] [answered]
I'm trying to understand how the library works and I'm stuck in a step. There is a generic class called `RasterDataset`, this class has a special method to get the items based on a bounding box, I assume that this is the method that is used when any of the Samplers wants to access to one of their items, right?

I want to use this package to train a multi-label classification model, to do so, I plan to create a `CustomDataSet` by inheriting from the `RasterDataset` class, and then override the special method `__getitem__` to return the `sample` dict with the `image` and the new `label` key... by doing this, as I've changed the way of accessing to the element (now by an index, instead of bbox), the Sampler classes won't be able to filter the dataset by this parameter, right?...

**Comment by adamjstewart** [comment, marked as answer]
Hi @dfguerrerom, thanks for the question. I think it may be helpful to read our [blog](https://pytorch.org/blog/geospatial-deep-learning-with-torchgeo/) or [paper](https://arxiv.org/abs/2111.08872) if you want a better idea of the design of the library.

TL;DR: In TorchGeo, there are two important base classes to know about:

* GeoDataset: for datasets with geospatial information, indexed using spatiotemporal bounding boxes, great for uncurated/raw data, allows you to combine multiple data sources based on geographic metadata, requires a GeoSampler to use.
* VisionDataset: for datasets without geospatial information, indexed by integer like in torchvision, great for curated datasets, can't combine with other datasets though, no sampler required.

From what you've described, it sounds like VisionDataset may be what you want. If you already have a curated dataset with both small image patches and labels, then there's no reason to use a GeoDataset. Note that RasterDataset is a subclass of GeoDataset that looks for raster files in a directory.

P.S. We're planning on renaming VisionDataset to NonGeoDataset in the near future, so the docs may change all of a sudden.

**Reply by dfguerrerom** [reply]
Hi @adamjstewart thanks for your answer.
Yes, it was a misunderstanding from my side because my dataset is composed of several `GeoTIFF` images, and they have geospatial information. Still, they are already curated to be used in a DL.
Perhaps reading a `VisionDataset` (`NonGeoDataset`) let me felt that my geographical dataset was not meant to be used there, but of course, we don't need the geographical info to perform any other task than a preprocessing step.

**Reply by adamjstewart** [reply]
Of course, it's still possible to use `GeoDataset` (`RasterDataset`) for your use case. You would build an r-tree index for either the images or labels, then override the `__getitem__` to read from both files instead of just the images or labels. You would probably want to use `PreChippedGeoSampler` for this dataset. The only advantage of using `GeoDataset` would be that you could potentially combine this dataset with another, or restrict it to a specific geospatial region.

**Comment by kimzed** [comment]
I am having the same issue. I have been trying torchgeo to see if it fits my needs, i am currently training a binary image classification model and i have a raster datasets. i want to find a way to get my classes using the RasterDataset, the same way for example a from torchvision.datasets import ImageFolder would do. but i do not find any implementation for image classification in the torchgeo library, and no similar case in torchgeo.datasets

**Reply by adamjstewart** [reply]
If you want something similar to `ImageFolder`, we have `NonGeoClassificationDataset` that behaves almost identically: https://torchgeo.readthedocs.io/en/stable/api/datasets.html#torchgeo.datasets.NonGeoClassificationDataset
